File: examples/thermal/mtpq/cpu_save.cpp
// =============================================================================
// examples/thermal/mtpq/cpu_save.cpp
//
// thermal | mTPQ | CPU (OpenMP) | full Hilbert (no symmetry)
// + persistent trajectory + state-vector snapshots at probe-betas (HDF5)
//
// Pillar 1 of the May 2026 "Save and DSSF Upgrades" plan. Demonstrates:
//
//     opts.output_dir  = "<dir>";
//     opts.probe_betas = {beta_1, beta_2, ...};
//
// causes `ed::api::thermal` (== `ed::workflows::thermal`) to persist:
//
//   * /tpq/samples/sample_<s>/thermodynamics  --
//        per-sample (beta, E, var, doublon, step) rows for the whole
//        mTPQ trajectory (one row per kernel step).
//   * /tpq/samples/sample_<s>/states/beta_<b> --
//        host-side state vector at the kernel-step beta closest to each
//        user-requested probe beta.
//
// `result.hdf5_path` reports the resulting file (mirrored into Python's
// `EDResults.eigenvectors_path` / `ThermalResult.hdf5_path`).
//
// The companion Python example reloads the persisted state via h5py and
// feeds it into `qed.spectral(method="GroundStateCF", initial_state=...)`
// for the TPQ-to-CF spectral pipeline; the C++ example below loads it
// via `HDF5IO::loadTPQState` and reports `<psi|psi> = 1` to verify the
// round-trip.
//
// Twin: examples/thermal/mtpq/cpu_save.py
//
// Requires: HDF5 (already required by the default CPU build)
// =============================================================================

#include "../../_shared/common.h"

#include <ed/api.h>
#include <ed/core/hdf5_io.h>

#include <cmath>
#include <complex>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <limits>
#include <vector>

int main(int argc, char** argv) {
    auto guard = ed_example::mpi_guard(argc, argv);
    (void)guard;
    constexpr std::uint64_t N = 8;

    auto op   = ed_example::heisenberg_chain(N, /*pbc=*/true);
    auto spec = ed_example::in_memory_spec(std::move(op), N);

    const std::string outdir = "ed_thermal_save_demo";
    std::filesystem::remove_all(outdir);

    ed::api::ThermalOptions opts;
    opts.method      = "mTPQ";
    opts.device      = "cpu";
    opts.T_min       = 0.1;
    opts.T_max       = 10.0;
    opts.num_T       = 8;
    opts.num_samples = 1;
    opts.random_seed = 42;
    opts.max_iterations          = 200;
    opts.use_sz_if_conserved     = false;
    opts.use_symmetry_if_available = false;
    opts.output_dir              = outdir;
    opts.probe_betas             = {1.0, 5.0};

    auto result = ed::api::thermal(std::move(spec), opts);

    std::cout << std::setprecision(6);
    ed_example::rank0_print("hdf5_path = ", result.hdf5_path, "\n");
    ed_example::rank0_print("gs_E      = ", result.ground_state_energy, "\n");

    // Round-trip: enumerate the snapshots persisted in the file and
    // load the one whose effective beta is closest to 5.0. The kernel
    // rounds each user-requested probe beta to the nearest
    // trajectory-step beta, so we ask for an exact beta match using
    // the value we discover in the HDF5 listing.
    auto snapshots = HDF5IO::listTPQStates(result.hdf5_path,
                                             /*sample_filter=*/0);
    double best_dist = std::numeric_limits<double>::infinity();
    double best_beta = 0.0;
    for (const auto& info : snapshots) {
        const double d = std::abs(info.beta - 5.0);
        if (d < best_dist) {
            best_dist = d;
            best_beta = info.beta;
        }
    }
    std::vector<std::complex<double>> psi_reload;
    const bool ok = HDF5IO::loadTPQState(result.hdf5_path,
                                          /*sample_index=*/0,
                                          best_beta,
                                          psi_reload);

    double norm2 = 0.0;
    for (const auto& z : psi_reload) {
        norm2 += std::norm(z);
    }
    ed_example::rank0_print("snapshot count           = ",
                             snapshots.size(), "\n");
    ed_example::rank0_print("closest beta to 5.0      = ",
                             std::setprecision(4), best_beta, "\n");
    ed_example::rank0_print("loaded |psi| dim         = ",
                             psi_reload.size(), "\n");
    ed_example::rank0_print("loaded <psi|psi>         = ",
                             std::setprecision(10), norm2, "\n");
    ed_example::rank0_print("HDF5IO::loadTPQState ok  = ", ok, "\n");

    // === Expected output (deterministic; captured on the CI reference runner) ===
    // hdf5_path = ed_thermal_save_demo/ed_results.h5
    // gs_E      = -3.6216
    // snapshot count           = 2
    // closest beta to 5.0      = 5.002
    // loaded |psi| dim         = 256
    // loaded <psi|psi>         = 1
    // HDF5IO::loadTPQState ok  = 1
    // ===========================================================================
    return 0;
}

File: examples/thermal/mtpq/cpu_save.py
"""thermal | mTPQ | CPU (OpenMP) | full Hilbert (no symmetry)
+ persistent trajectory + state-vector snapshots at probe-betas (HDF5)

Pillar 1 of the May 2026 "Save and DSSF Upgrades" plan. Demonstrates::

    qed.thermal(H, ..., method="mTPQ",
                output_dir="<dir>",
                probe_betas=[beta_1, beta_2, ...])

writes per-sample trajectories and host-side state vectors at the
kernel-step beta closest to each user-requested probe beta into
``<dir>/ed_results.h5`` and surfaces the path via ``result.hdf5_path``.

The state snapshots land at ``/tpq/samples/sample_<s>/states/beta_<b>``
and can be reloaded via h5py for the TPQ-to-CF spectral pipeline (see
``examples/spectral/ground_state_dssf/cpu_from_tpq_state.py``).

Twin: ``examples/thermal/mtpq/cpu_save.cpp``.

Requires: ``h5py``.
"""
from __future__ import annotations

import os
import shutil
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from _shared.common import heisenberg_chain, rank0_print

import h5py
import numpy as np
import qed

N = 8
H = heisenberg_chain(N, pbc=True)

outdir = "ed_thermal_save_demo"
if os.path.isdir(outdir):
    shutil.rmtree(outdir)

result = qed.thermal(
    H,
    method="mTPQ",
    T_min=0.1,
    T_max=10.0,
    num_T=8,
    num_samples=1,
    random_seed=42,
    max_iterations=200,
    device="cpu",
    output_dir=outdir,
    probe_betas=[1.0, 5.0],
    use_sz_if_conserved=False,
    verbose=False,
)

rank0_print(f"hdf5_path = {result.hdf5_path}")
rank0_print(f"gs_E      = {result.ground_state_energy:.6f}")

# Reload via h5py to verify the on-disk artefact carries the expected
# structure: per-sample trajectory rows + state vectors at the
# kernel-step betas closest to the user-requested probe betas.
with h5py.File(result.hdf5_path, "r") as f:
    sample0 = f["/tpq/samples/sample_0"]
    traj    = sample0["thermodynamics"][...]  # (n_steps, 5)
    states  = list(sample0["states"].keys())
    rank0_print(f"trajectory rows         = {traj.shape[0]}")
    rank0_print(f"snapshot datasets       = {sorted(states)}")
    psi5_name = sorted(states, key=lambda k: abs(
        float(k.removeprefix("beta_")) - 5.0))[0]
    # The HDF5 layout uses a compound (real, imag) datatype; collapse to
    # std::complex<double> on the Python side.
    raw = np.asarray(sample0["states"][psi5_name][...])
    psi5 = raw["real"] + 1j * raw["imag"]
    norm2 = float(np.vdot(psi5, psi5).real)
    rank0_print(f"|psi(beta~=5)|.shape    = {psi5.shape}")
    rank0_print(f"<psi|psi>               = {norm2:.6f}")

# === Expected output (deterministic; captured on the CI reference runner) ===
# hdf5_path = ed_thermal_save_demo/ed_results.h5
# gs_E      = -3.637107
# trajectory rows         = 201
# snapshot datasets       = ['beta_1.004994', 'beta_5.001079']
# |psi(beta~=5)|.shape    = (256,)
# <psi|psi>               = 1.000000
# ===========================================================================
